api/client/teleport_resolver.go

```diff
@@ -0,0 +1,173 @@
+/*
+ * Teleport
+ * Copyright (C) 2025  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package client
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/gravitational/teleport/api/client/webclient"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/serviceconfig"
+)
+
+// teleportResolverBuilder is responsible for building a wrapped DNS resolver that polls the provided
+// PingURL to determine what the service configuration should be.
+//
+// teleportResolverBuilder is intentionally not registered as a [resolver.Builder] so that gRPC clients
+// must explicitly opt in to using a wrapped DNS resolver.
+type teleportResolverBuilder struct {
+	PingURL string
+}
+
+// Build returns a wrapped DNS resolver that polls the configured PingURL to determine what the
+// service configuration should be.
+func (r teleportResolverBuilder) Build(target resolver.Target, cc resolver.ClientConn, opts resolver.BuildOptions) (resolver.Resolver, error) {
+	dnsBuilder := resolver.Get("dns")
+	if dnsBuilder == nil {
+		return nil, errors.New("unable to get dns resolver")
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	wr := wrappedResolver{
+		ClientConn: cc,
+
+		ctx:    ctx,
+		cancel: cancel,
+	}
+
+	dnsResolver, err := dnsBuilder.Build(target, &wr, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	wr.Resolver = dnsResolver
+
+	log := slog.New(slog.NewJSONHandler(os.Stderr, nil))
+
+	go func() {
+		for {
+			select {
+			case <-time.After(30 * time.Second):
```

Comment:
Decide on an interval. Also add support for debouncing to prevent flapping back and forth during a proxy rollout.